README
NAME
    Log::Any - Bringing loggers and listeners together

VERSION
    version 1.03

SYNOPSIS
    In a CPAN or other module:

        package Foo;
        use Log::Any qw($log);

        # log a string
        $log->error("an error occurred");

        # log a string and data using a formatting filter
        $log->debugf("arguments are: %s", \@_);

    In a Moo/Moose-based module:

        package Foo;
        use Moo;

        has log => (
            is => 'ro',
            isa => 'Log::Any::Proxy',
            default => sub { Log::Any->get_logger },
        );

    In your application:

        use Foo;
        use Log::Any::Adapter;

        # Send all logs to Log::Log4perl
        Log::Any::Adapter->set('Log4perl');

        # Send all logs to Log::Dispatch
        my $log = Log::Dispatch->new(outputs => [[ ... ]]);
        Log::Any::Adapter->set( 'Dispatch', dispatcher => $log );

        # See Log::Any::Adapter documentation for more options

DESCRIPTION
    "Log::Any" provides a standard log production API for modules.
    Log::Any::Adapter allows applications to choose the mechanism for log
    consumption, whether screen, file or another logging mechanism like
    Log::Dispatch or Log::Log4perl.

    Many modules have something interesting to say. Unfortunately there is
    no standard way for them to say it - some output to STDERR, others to
    "warn", others to custom file logs. And there is no standard way to get
    a module to start talking - sometimes you must call a uniquely named
    method, other times set a package variable.

    This being Perl, there are many logging mechanisms available on CPAN.
    Each has their pros and cons. Unfortunately, the existence of so many
    mechanisms makes it difficult for a CPAN author to commit his/her users
    to one of them. This may be why many CPAN modules invent their own
    logging or choose not to log at all.

    To untangle this situation, we must separate the two parts of a logging
    API. The first, *log production*, includes methods to output logs (like
    "$log->debug") and methods to inspect whether a log level is activated
    (like "$log->is_debug"). This is generally all that CPAN modules care
    about. The second, *log consumption*, includes a way to configure where
    logging goes (a file, the screen, etc.) and the code to send it there.
    This choice generally belongs to the application.

    A CPAN module uses "Log::Any" to get a log producer object. An
    application, in turn, may choose one or more logging mechanisms via
    Log::Any::Adapter, or none at all.

    "Log::Any" has a very tiny footprint and no dependencies beyond Perl
    5.8.1, which makes it appropriate for even small CPAN modules to use. It
    defaults to 'null' logging activity, so a module can safely log without
    worrying about whether the application has chosen (or will ever choose)
    a logging mechanism.

    See <http://www.openswartz.com/2007/09/06/standard-logging-api/> for the
    original post proposing this module.

LOG LEVELS
    "Log::Any" supports the following log levels and aliases, which is meant
    to be inclusive of the major logging packages:

         trace
         debug
         info (inform)
         notice
         warning (warn)
         error (err)
         critical (crit, fatal)
         alert
         emergency

    Levels are translated as appropriate to the underlying logging
    mechanism. For example, log4perl only has six levels, so we translate
    'notice' to 'info' and the top three levels to 'fatal'. See the
    documentation of an adapter class for specifics.

CATEGORIES
    Every logger has a category, generally the name of the class that asked
    for the logger. Some logging mechanisms, like log4perl, can direct logs
    to different places depending on category.

PRODUCING LOGS (FOR MODULES)
  Getting a logger
    The most convenient way to get a logger in your module is:

        use Log::Any qw($log);

    This creates a package variable *$log* and assigns it to the logger for
    the current package. It is equivalent to

        our $log = Log::Any->get_logger;

    In general, to get a logger for a specified category:

        my $log = Log::Any->get_logger(category => $category)

    If no category is specified, the calling package is used.

    A logger object is an instance of Log::Any::Proxy, which passes on
    messages to the Log::Any::Adapter handling its category.

  Logging
    To log a message, pass a single string to any of the log levels or
    aliases. e.g.

        $log->error("this is an error");
        $log->warn("this is a warning");
        $log->warning("this is also a warning");

    You should not include a newline in your message; that is the
    responsibility of the logging mechanism, which may or may not want the
    newline.

    There are also versions of each of these methods with an additional "f"
    suffix ("infof", "errorf", "debugf", etc.) that format a list of
    arguments. The specific formatting mechanism and meaning of the
    arguments is controlled by the Log::Any::Proxy object.

        $log->errorf("an error occurred: %s", $@);
        $log->debugf("called with %d params: %s", $param_count, \@params);

    By default it renders like "sprintf", with the following additional
    features:

    *   Any complex references (like "\@params" above) are automatically
        converted to single-line strings with "Data::Dumper".

    *   Any undefined values are automatically converted to the string
        "<undef>".

  Log level detection
    To detect whether a log level is on, use "is_" followed by any of the
    log levels or aliases. e.g.

        if ($log->is_info()) { ... }
        $log->debug("arguments are: " . Dumper(\@_))
            if $log->is_debug();

    This is important for efficiency, as you can avoid the work of putting
    together the logging message (in the above case, stringifying @_) if the
    log level is not active.

    The formatting methods ("infof", "errorf", etc.) check the log level for
    you.

    Some logging mechanisms don't support detection of log levels. In these
    cases the detection methods will always return 1.

    In contrast, the default logging mechanism - Null - will return 0 for
    all detection methods.

  Setting an alternate default logger
    To choose something other than Null as the default, pass it as a
    parameter when loading "Log::Any"

        use Log::Any '$log', default_adapter => 'Stderr';

    The name of the default class follows the same rules as used by
    Log::Any::Adapter.

  Configuring the proxy
    Any parameter passed on the import line or via the "get_logger" method
    are passed on the the Log::Any::Proxy constructor.

        use Log::Any '$log', filter => \&myfilter;

  Testing
    Log::Any::Test provides a mechanism to test code that uses "Log::Any".

CONSUMING LOGS (FOR APPLICATIONS)
    Log::Any provides modules with a Log::Any::Proxy object, which is the
    log producer. To consume its output and direct it where you want (a
    file, the screen, syslog, etc.), you use Log::Any::Adapter along with a
    destination-specific subclass.

    For example, to send output to a file via Log::Any::Adapter::File, your
    application could do this:

        use Log::Any::Adapter ('File', '/path/to/file.log');

    See the Log::Any::Adapter documentation for more details.

Q & A
    Isn't Log::Any just yet another logging mechanism?
        No. "Log::Any" does not include code that knows how to log to a
        particular place (file, screen, etc.) It can only forward logging
        requests to another logging mechanism.

    Why don't you just pick the best logging mechanism, and use and promote
    it?
        Each of the logging mechanisms have their pros and cons,
        particularly in terms of how they are configured. For example,
        log4perl offers a great deal of power and flexibility but uses a
        global and potentially heavy configuration, whereas "Log::Dispatch"
        is extremely configuration-light but doesn't handle categories.
        There is also the unnamed future logger that may have advantages
        over either of these two, and all the custom in-house loggers people
        have created and cannot (for whatever reason) stop using.

    Is it safe for my critical module to depend on Log::Any?
        Our intent is to keep "Log::Any" minimal, and change it only when
        absolutely necessary. Most of the "innovation", if any, is expected
        to occur in "Log::Any::Adapter", which your module should not have
        to depend on (unless it wants to direct logs somewhere specific).
        "Log::Any" has no non-core dependencies.

    Why doesn't Log::Any use *insert modern Perl technique*?
        To encourage CPAN module authors to adopt and use "Log::Any", we aim
        to have as few dependencies and chances of breakage as possible.
        Thus, no "Moose" or other niceties.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/Log-Any/issues>. You will be notified
    automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/Log-Any>

      git clone https://github.com/dagolden/Log-Any.git

AUTHORS
    *   Jonathan Swartz <swartz@pobox.com>

    *   David Golden <dagolden@cpan.org>

CONTRIBUTORS
    *   Maxim Vuets <maxim.vuets@booking.com>

    *   Stephen Thirlwall <sdt@dr.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Jonathan Swartz and David Golden.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Log-Any

** denotes an incompatible change

1.03      2015-01-01 22:39:41-05:00 America/New_York

    [Changed]

    - Log::Any::Proxy concatenates arguments to basic logging functions
      with a space character before passing them to adapters as a single
      string.  This ensures consistency across adapters that handle
      multiple arguments differently.

1.02      2014-12-28 07:06:49-05:00 America/New_York

    [Fixed]

    - Some adapters relied on Log::Any::Adapter::Util also loading
      Log::Any so this behavior has been restored.

1.01      2014-12-26 22:25:13-05:00 America/New_York

    [Fixed]

    - 'numeric_level' was not exported properly from
      Log::Any::Adapter::Util

1.00      2014-12-25 22:04:13-05:00 America/New_York

    [Added]

    - Logging now goes via a Log::Any::Proxy object instead of directly to
      an adapter. This allows easy customization of the message production.

    - File, Stdout, and Stderr adapters now support a minimum
      log level parameter.

    [Changed]

    - Removed dead code from Log::Any::Adapter::Base; particularly this
      was the formatting code, since this is now handled by
      Log::Any::Proxy.

    [Fixed]

    - File will flock the handle when writing (if flock is avaiable).

    - Won't die if adapters aren't loadable modules as long as they
      provide a constructor.  This allows using private adapters
      defined in another file.

    [Documented]

    - Revised docs for creating adapters

    - Fixed typos and improved docs for Log::Any::Adapter::Util; removed
      stub docs for modules that didn't need it.

    [Deprecated]

    - Deprecated some methods in Log::Any::Adapter::Util

    [Internal]

    - Merged Log-Any and Log-Any-Adapter distributions; reduces code
      duplication and ensures Log::Any and adapter framework stay in sync

    - Eliminates all non-core dependencies (as of Perl 5.8.1), including
      Capture::Tiny, Devel::GlobalDestruction and Guard

0.92      2014-12-15 07:12:38-05:00 America/New_York (TRIAL RELEASE)

0.91      2014-12-14 22:13:09-05:00 America/New_York (TRIAL RELEASE)

0.90      2014-12-12 17:08:22-05:00 America/New_York (TRIAL RELEASE)

0.15  Apr 10, 2013

* Fixes
  - Hide 'package Log::Any::Adapter' from PAUSE/Module::Metadata - miyagawa

0.14  Aug 31, 2011

* Fixes
  - Fix version number in Log/Any.pm - Stephen Thirlwall

0.13  Aug 2, 2011

* Fixes
  - Fix typo in lib/Log/Any/Adapter/Test.pm - RT #69850 - Stephen Thirlwall

0.12  Mar 23, 2011

* Fixes
  - Return false from null adapter is_xxx methods - RT #64164 - Chip Salzenberg
  - Eliminate 'subroutine redefined' warning in case Log::Any::Adapter loaded before
    Log::Any::Test

* Implementation
  - Migrate to Dist::Zilla

0.11  Feb 12, 2010

* Improvements
  - Add trace level - suggested by Szymon Swierkosz

0.10  Jan 5, 2010

* Fixes
  - Fix Log::Any::Core to support references in printf-style methods

0.09  Jan 5, 2010

* Improvements
  - Convert undef to string "<undef>" in printf-style methods - RT #53398, suggested by Steven Haryanto

0.08  Dec 15, 2009

* Fixes
  - Fix Log::Any::Test to support full logging API (aliases and printf methods)

0.07  Dec 7, 2009

* Implementation
  - Depend on Test::Simple rather than Test::More

* Improvements
  - Add Log::Any::Test, to aid in testing code that uses Log::Any

0.06  Oct 31, 2009

* Fixes
  - Implement Log::Any->set_adapter again for backward compatibility with 0.04 and earlier

0.05  Oct 27, 2009

* Implementation
  - ** Strip Log::Any down to a relative minimum, so as to keep it stable and
    unchanging. Move everything else to Log::Any::Adapter.

0.04  Sep 3, 2009

* Fixes
  - Replace Test::Deep::cmp_deeply with an internal version to avoid Test::Deep dependency

0.03  Jul 17, 2009

* Improvements
  - Eliminated extra '::Log' from adapter names, e.g. Log::Any::Adapter::Log::Dispatch is
    now Log::Any::Adapter::Dispatch. The long name was properly descriptive but was making
    me wince every time.

0.02  Jul 14, 2009

* Fixes
  - Fix logging aliases like warn => warning

* Implementation
  - Eliminate unnecessary Test/InternalOnly.pm class
  - Precompute alias and method lists like Log::Any->logging_methods

0.01  Jul 11, 2009

- Initial version



lib/Log/Any.pm
use 5.008001;
use strict;
use warnings;

package Log::Any;

# ABSTRACT: Bringing loggers and listeners together
our $VERSION = '1.03';

use Carp ();
use Log::Any::Manager;
use Log::Any::Adapter::Util qw(
  require_dynamic
  detection_aliases
  detection_methods
  log_level_aliases
  logging_aliases
  logging_and_detection_methods
  logging_methods
);

# This is overridden in Log::Any::Test
our $OverrideDefaultAdapterClass;
our $OverrideDefaultProxyClass;

# singleton and accessor
{
    my $manager = Log::Any::Manager->new();
    sub _manager { return $manager }
}

sub import {
    my $class  = shift;
    my $caller = caller();

    my @export_params = ( $caller, @_ );
    $class->_export_to_caller(@export_params);
}

sub _export_to_caller {
    my $class  = shift;
    my $caller = shift;

    # Parse parameters passed to 'use Log::Any'
    my $saw_log_param;
    my @params;
    while ( my $param = shift @_ ) {
        if ( $param eq '$log' ) {
            $saw_log_param = 1;    # defer until later
            next;                  # singular
        }
        else {
            push @params, $param, shift @_;    # pairwise
        }
    }

    Carp::croak("Argument list not balanced: @params")
      unless @params % 2 == 0;

    # get logger if one was requested
    if ($saw_log_param) {
        no strict 'refs';
        my $proxy = $class->get_logger( category => $caller, @params );
        my $varname = "$caller\::log";
        *$varname = \$proxy;
    }
}

sub get_logger {
    my ( $class, %params ) = @_;
    no warnings 'once';

    my $proxy_class = $class->_get_proxy_class( delete $params{proxy_class} );
    my $category =
      defined $params{category} ? delete $params{'category'} : caller;

    if ( my $default = delete $params{'default_adapter'} ) {
        $class->_manager->set_default( $category, $default );
    }

    my $adapter = $class->_manager->get_adapter( $category );

    require_dynamic($proxy_class);
    return $proxy_class->new(
        %params, adapter => $adapter, category => $category,
    );
}

sub _get_proxy_class {
    my ( $self, $proxy_name ) = @_;
    return $Log::Any::OverrideDefaultProxyClass
      if $Log::Any::OverrideDefaultProxyClass;
    return "Log::Any::Proxy" unless $proxy_name;
    my $proxy_class = (
          substr( $proxy_name, 0, 1 ) eq '+'
        ? substr( $proxy_name, 1 )
        : "Log::Any::Proxy::$proxy_name"
    );
    return $proxy_class;
}

# For backward compatibility
sub set_adapter {
    my $class = shift;
    Log::Any->_manager->set(@_);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Any - Bringing loggers and listeners together

=head1 VERSION

version 1.03

=head1 SYNOPSIS

In a CPAN or other module:

    package Foo;
    use Log::Any qw($log);

    # log a string
    $log->error("an error occurred");

    # log a string and data using a formatting filter
    $log->debugf("arguments are: %s", \@_);

In a Moo/Moose-based module:

    package Foo;
    use Moo;

    has log => (
        is => 'ro',
        isa => 'Log::Any::Proxy',
        default => sub { Log::Any->get_logger },
    );

In your application:

    use Foo;
    use Log::Any::Adapter;

    # Send all logs to Log::Log4perl
    Log::Any::Adapter->set('Log4perl');

    # Send all logs to Log::Dispatch
    my $log = Log::Dispatch->new(outputs => [[ ... ]]);
    Log::Any::Adapter->set( 'Dispatch', dispatcher => $log );

    # See Log::Any::Adapter documentation for more options

=head1 DESCRIPTION

C<Log::Any> provides a standard log production API for modules.
L<Log::Any::Adapter> allows applications to choose the mechanism for log
consumption, whether screen, file or another logging mechanism like
L<Log::Dispatch> or L<Log::Log4perl>.

Many modules have something interesting to say. Unfortunately there is no
standard way for them to say it - some output to STDERR, others to C<warn>,
others to custom file logs. And there is no standard way to get a module to
start talking - sometimes you must call a uniquely named method, other times
set a package variable.

This being Perl, there are many logging mechanisms available on CPAN.  Each has
their pros and cons. Unfortunately, the existence of so many mechanisms makes
it difficult for a CPAN author to commit his/her users to one of them. This may
be why many CPAN modules invent their own logging or choose not to log at all.

To untangle this situation, we must separate the two parts of a logging API.
The first, I<log production>, includes methods to output logs (like
C<$log-E<gt>debug>) and methods to inspect whether a log level is activated
(like C<$log-E<gt>is_debug>). This is generally all that CPAN modules care
about. The second, I<log consumption>, includes a way to configure where
logging goes (a file, the screen, etc.) and the code to send it there. This
choice generally belongs to the application.

A CPAN module uses C<Log::Any> to get a log producer object.  An application,
in turn, may choose one or more logging mechanisms via L<Log::Any::Adapter>, or
none at all.

C<Log::Any> has a very tiny footprint and no dependencies beyond Perl 5.8.1,
which makes it appropriate for even small CPAN modules to use. It defaults to
'null' logging activity, so a module can safely log without worrying about
whether the application has chosen (or will ever choose) a logging mechanism.

See L<http://www.openswartz.com/2007/09/06/standard-logging-api/> for the
original post proposing this module.

=head1 LOG LEVELS

C<Log::Any> supports the following log levels and aliases, which is meant to be
inclusive of the major logging packages:

     trace
     debug
     info (inform)
     notice
     warning (warn)
     error (err)
     critical (crit, fatal)
     alert
     emergency

Levels are translated as appropriate to the underlying logging mechanism. For
example, log4perl only has six levels, so we translate 'notice' to 'info' and
the top three levels to 'fatal'.  See the documentation of an adapter class
for specifics.

=head1 CATEGORIES

Every logger has a category, generally the name of the class that asked for the
logger. Some logging mechanisms, like log4perl, can direct logs to different
places depending on category.

=head1 PRODUCING LOGS (FOR MODULES)

=head2 Getting a logger

The most convenient way to get a logger in your module is:

    use Log::Any qw($log);

This creates a package variable I<$log> and assigns it to the logger for the
current package. It is equivalent to

    our $log = Log::Any->get_logger;

In general, to get a logger for a specified category:

    my $log = Log::Any->get_logger(category => $category)

If no category is specified, the calling package is used.

A logger object is an instance of L<Log::Any::Proxy>, which passes
on messages to the L<Log::Any::Adapter> handling its category.

=head2 Logging

To log a message, pass a single string to any of the log levels or aliases. e.g.

    $log->error("this is an error");
    $log->warn("this is a warning");
    $log->warning("this is also a warning");

You should B<not> include a newline in your message; that is the responsibility
of the logging mechanism, which may or may not want the newline.

There are also versions of each of these methods with an additional "f" suffix
(C<infof>, C<errorf>, C<debugf>, etc.) that format a list of arguments.  The
specific formatting mechanism and meaning of the arguments is controlled by the
L<Log::Any::Proxy> object.

    $log->errorf("an error occurred: %s", $@);
    $log->debugf("called with %d params: %s", $param_count, \@params);

By default it renders like C<sprintf>, with the following additional features:

=over

=item *

Any complex references (like C<\@params> above) are automatically converted to
single-line strings with C<Data::Dumper>.

=item *

Any undefined values are automatically converted to the string "<undef>".

=back

=head2 Log level detection

To detect whether a log level is on, use "is_" followed by any of the log
levels or aliases. e.g.

    if ($log->is_info()) { ... }
    $log->debug("arguments are: " . Dumper(\@_))
        if $log->is_debug();

This is important for efficiency, as you can avoid the work of putting together
the logging message (in the above case, stringifying C<@_>) if the log level is
not active.

The formatting methods (C<infof>, C<errorf>, etc.) check the log level for you.

Some logging mechanisms don't support detection of log levels. In these cases
the detection methods will always return 1.

In contrast, the default logging mechanism - Null - will return 0 for all
detection methods.

=head2 Setting an alternate default logger

To choose something other than Null as the default, pass it as a parameter when
loading C<Log::Any>

    use Log::Any '$log', default_adapter => 'Stderr';

The name of the default class follows the same rules as used by L<Log::Any::Adapter>.

=head2 Configuring the proxy

Any parameter passed on the import line or via the C<get_logger> method
are passed on the the L<Log::Any::Proxy> constructor.

    use Log::Any '$log', filter => \&myfilter;

=head2 Testing

L<Log::Any::Test> provides a mechanism to test code that uses C<Log::Any>.

=head1 CONSUMING LOGS (FOR APPLICATIONS)

Log::Any provides modules with a L<Log::Any::Proxy> object, which is the log
producer.  To consume its output and direct it where you want (a file, the
screen, syslog, etc.), you use L<Log::Any::Adapter> along with a
destination-specific subclass.

For example, to send output to a file via L<Log::Any::Adapter::File>, your
application could do this:

    use Log::Any::Adapter ('File', '/path/to/file.log');

See the L<Log::Any::Adapter> documentation for more details.

=head1 Q & A

=over

=item Isn't Log::Any just yet another logging mechanism?

No. C<Log::Any> does not include code that knows how to log to a particular
place (file, screen, etc.) It can only forward logging requests to another
logging mechanism.

=item Why don't you just pick the best logging mechanism, and use and promote it?

Each of the logging mechanisms have their pros and cons, particularly in terms
of how they are configured. For example, log4perl offers a great deal of power
and flexibility but uses a global and potentially heavy configuration, whereas
C<Log::Dispatch> is extremely configuration-light but doesn't handle
categories. There is also the unnamed future logger that may have advantages
over either of these two, and all the custom in-house loggers people have
created and cannot (for whatever reason) stop using.

=item Is it safe for my critical module to depend on Log::Any?

Our intent is to keep C<Log::Any> minimal, and change it only when absolutely
necessary. Most of the "innovation", if any, is expected to occur in
C<Log::Any::Adapter>, which your module should not have to depend on (unless it
wants to direct logs somewhere specific). C<Log::Any> has no non-core dependencies.

=item Why doesn't Log::Any use I<insert modern Perl technique>?

To encourage CPAN module authors to adopt and use C<Log::Any>, we aim to have
as few dependencies and chances of breakage as possible. Thus, no C<Moose> or
other niceties.

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/Log-Any/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/Log-Any>

  git clone https://github.com/dagolden/Log-Any.git

=head1 AUTHORS

=over 4

=item *

Jonathan Swartz <swartz@pobox.com>

=item *

David Golden <dagolden@cpan.org>

=back

=head1 CONTRIBUTORS

=for stopwords Maxim Vuets Stephen Thirlwall

=over 4

=item *

Maxim Vuets <maxim.vuets@booking.com>

=item *

Stephen Thirlwall <sdt@dr.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Jonathan Swartz and David Golden.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
